flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/rules/logical/FlinkLogicalRankRule.scala

```diff
@@ -0,0 +1,231 @@
+
+import scala.collection.JavaConversions._
+
+/**
+  * Planner rule that matches a [[FlinkLogicalCalc]] on a [[FlinkLogicalOverWindow]],
+  * and converts them into a [[FlinkLogicalRank]].
+  */
+abstract class FlinkLogicalRankRuleBase
+  extends RelOptRule(
+    operand(classOf[FlinkLogicalCalc],
+      operand(classOf[FlinkLogicalOverWindow], any()))) {
+
+  override def onMatch(call: RelOptRuleCall): Unit = {
+    val calc: FlinkLogicalCalc = call.rel(0)
+    val window: FlinkLogicalOverWindow = call.rel(1)
+    val group = window.groups.get(0)
+    val rankFun = group.aggCalls.get(0).getOperator.asInstanceOf[SqlRankFunction]
+
+    // the rank function is the last field of LogicalWindow
+    val rankFieldIndex = window.getRowType.getFieldCount - 1
+    val condition = calc.getProgram.getCondition
+    val predicate = calc.getProgram.expandLocalRef(condition)
+
+    val config = calc.getCluster.getPlanner.getContext.asInstanceOf[FlinkContext].getTableConfig
+    val (rankRange, remainingPreds) = FlinkRelOptUtil.extractRankRange(
+      predicate,
+      rankFieldIndex,
+      calc.getCluster.getRexBuilder,
+      config)
+    require(rankRange.isDefined)
+
+    val cluster = window.getCluster
+    val rexBuilder = cluster.getRexBuilder
+
+    val calcProgram = calc.getProgram
+    val exprList = calcProgram.getProjectList.map(calcProgram.expandLocalRef)
+
+    val inputFields = RelOptUtil.InputFinder.bits(exprList, null).toList
+    val outputRankFunColumn = inputFields.contains(rankFieldIndex)
+
+    val rankRowType = if (outputRankFunColumn) {
+      window.getRowType
+    } else {
+      val typeBuilder = rexBuilder.getTypeFactory.builder()
+      window.getRowType.getFieldList.dropRight(1).foreach(typeBuilder.add)
+      typeBuilder.build()
+    }
+
+    rankRange match {
+      case Some(ConstantRankRange(_, rankEnd)) if rankEnd <= 0 =>
+        throw new TableException(s"Rank end should not less than zero, but now is $rankEnd")
+      case _ => // do nothing
+    }
+
+    val rank = new FlinkLogicalRank(
+      cluster,
+      window.getTraitSet,
+      window.getInput,
+      rankFun,
+      group.keys,
+      group.orderKeys,
+      rankRange.get,
+      outputRankFunColumn)
+
+    val newRel = if (RexUtil.isIdentity(exprList, rankRowType) && remainingPreds.isEmpty) {
+      // project is trivial and filter is empty, remove the Calc
+      rank
+    } else {
+      val programBuilder = RexProgramBuilder.create(
+        rexBuilder,
+        rankRowType,
+        calcProgram.getExprList,
+        calcProgram.getProjectList,
+        remainingPreds.orNull,
+        calc.getRowType,
+        true, // normalize
+        null) // simplify
+
+      calc.copy(calc.getTraitSet, rank, programBuilder.getProgram)
+    }
+    call.transformTo(newRel)
+  }
+}
+
+/**
+  * This rule handles RANK function and rank range with end.
+  */
+class FlinkLogicalRankRuleForRangeEnd extends FlinkLogicalRankRuleBase {
```

Comment:
Can you give some examples to show what this rule want to capture?

Reply:
ok